### DISCUSSION TITLE
ly2 schema node key question

### DISCUSSION THREAD

**Discussion by choppsv1** [CESNET/libyang#1396] [Q&A]
I've written a macro to walk the compiled schema node list of keys.

```
#define LY_FOR_KEYS(snode, skey)                                           \
    for ((skey) = (const struct lysc_node_leaf *)lysc_node_child((snode)); \
        (skey);                                                            \
        (skey) = (const struct lysc_node_leaf *)((skey)->next))            \
            if (!(((skey)->nodetype & LYS_LEAF) && ((skey)->flags & LYS_KEY))) \
                continue;                                                  \
            else
```

Is it the case though that keys will always be at the start of the child list and so a `break` works better here than the `continue`?

**Comment by michalvasko** [comment]
Yes, keys will always be the first children of a list (even if written in different order in YANG). Also, you may want to use `lysc_is_key()`.

**Comment by choppsv1** [comment]
Revisiting this as we are counting on this in a few places in the code now, and with the recent confusion on what an API is and backward compatibility/semver/etc -- Is this particular trait part of the official API, and if so might it be documented as such?

Would be great if so, otherwise maybe we can't count on it...

**Reply by michalvasko** [reply]
Currently, I see no reason to ever change this as it is convenient but you are right, theoretically it can still change in the future. For example, if yang-next puts some meaning to the order of the schema nodes in YANG modules. So, I would prefer to keep everything as it is. There is the flag `LYS_KEY` that marks the list keys so use that. I am sure you can come up with code that would find the keys efficiently if they are the first schema children but at the same time handle situations if they are not.
